finetuner/tuner/evaluation.py

```diff
@@ -0,0 +1,160 @@
+from ..embedding import embed
+
+if TYPE_CHECKING:
+    from ..helper import AnyDNN, DocumentSequence
+
+
+METRICS = {
+    "r_precision": r_precision,
+    "precision": precision_at_k,
+    "recall": recall_at_k,
+    "f1score": f1_score_at_k,
+    "average_precision": average_precision,
+    "hit": hit_at_k,
+    "reciprocal_rank": reciprocal_rank,
+    "dcg": dcg_at_k,
+    "ndcg": ndcg_at_k,
+}
+
+
+class Evaluator:
+    def __init__(
+        self,
+        eval_data: 'DocumentSequence',
+        catalog: 'DocumentSequence',
+        embed_model: Optional['AnyDNN'] = None,
+        distance_metric: str = 'cosine',
+        limit: int = 20,
+    ):
+        """
+        Build an Evaluator object that can be used to evaluate an embedding model on a retrieval task
+
+        :param eval_data: a sequence of documents. Each document should contain ground truth matches under
+            ``doc.matches`` and relevance scores under ``doc.tags[__default_tag_key__]['label']``
+        :param catalog: a sequence of documents, against whist the eval docs will be matched.
+        :param embed_model: the embedding model to use, in order to extract document representations.
+        :param distance_metric: which distance metric to use when matching documents
+        :param limit: limit the number of results during matching.
+        :return: None.
+        """
+        self._embed_model = embed_model
+        self._eval_data = eval_data
+        self._catalog = catalog
+        self._summary_docs = self._parse_eval_docs()
+        self._distance_metric = distance_metric
+        self._limit = limit
+
+    @staticmethod
+    def _doc_to_relevance(doc: Document) -> Tuple[List[int], int]:
+        """
+        Convert a Jina document to a relevance representation
+        :param doc: A Jina document. The matched document identifiers, predicted by the search system, are expected to
+            be under ``doc.matches``. The expected documents identifiers along with the relevance scores, given by the
+            user as matching ground truth, should be under ``doc.tags[finetuner.__default_tag_key__]['targets']``.
+        :return: the relevance vector and the total number of relevant documents.
+        """
+        targets: Dict[str, int] = doc.tags[__default_tag_key__]['targets']
+        return [
+            targets[match.id] if match.id in targets else 0 for match in doc.matches
+        ], len(targets)
+
+    def _parse_eval_docs(self) -> DocumentArray:
+        """
+        Convert the evaluation docs to the internal representation used by the Evaluator
+        """
+        to_be_scored_docs = DocumentArray()
+        for doc in self._eval_data:
+            relevancies = [
+                (m.id, m.tags[__default_tag_key__]['label'])
+                for m in doc.matches
+                if m.tags[__default_tag_key__]['label'] > 0
+            ]
+            sorted_relevancies = sorted(relevancies, key=lambda x: x[1])
```

Comment:
no need to change, but just to let you know, we allow something like:

```python

matches = doc.matches.sort(key=lambda d: d.tags['id'], reverse=True)
for d in filter(lambda d: d.tags[__default_tag_key__]['label'] > 0, doc.matches):
    print(d)
```

checkout the docs [here](https://docs.jina.ai/fundamentals/document/documentarray/?highlight=filter#filter)

Reply:
This changed recently, now `__default_tag_key__='finetuner_label'`, so I think no need for `'label'`.

Reply:
changed to `[__default_tag_key__]` instead of `[__default_tag_key__]['label']`